backend/backend/routes/user_actions.py

```diff
@@ -0,0 +1,178 @@
+def remove_companies_from_collection(
+    request: RemoveCompaniesFromCollectionRequest,
+    db: Session = Depends(database.get_db)
+):
+    collection = get_collection_by_id(db, request.collection_id)
+    associations = get_existing_associations(db, request.company_ids, collection.id)
+    
+    if not associations:
+        raise HTTPException(status_code=400, detail="No matching company associations found")
+    
+    for assoc in associations:
+        db.delete(assoc)
+    db.commit()
+    
+    return {"message": f"{len(associations)} companies removed from '{collection.collection_name}' collection successfully"}
+
+@router.post("/like-company")
+def like_company(
+    request: LikeCompanyRequest,
+    db: Session = Depends(database.get_db)
+) -> LikedCompaniesResponse:
+    # Assuming there is a Liked Companies list
+    liked_collection = (
+        db.query(database.CompanyCollection)
+        .filter(database.CompanyCollection.collection_name == "Liked Companies")
+        .first()
+    )
+
+    # Check if company already in liked collection
+    liked_association = db.query(database.CompanyCollectionAssociation).filter(
+        database.CompanyCollectionAssociation.collection_id == liked_collection.id,
+        database.CompanyCollectionAssociation.company_id == request.company_id
+    ).first()
+    if liked_association:
+        return LikedCompaniesResponse(message="Company already liked", success=False)
+    
+    # Add company to liked collection
+    liked_association = database.CompanyCollectionAssociation(company_id=request.company_id, collection_id=liked_collection.id)
+    db.add(liked_association)
+    db.commit()
+    return LikedCompaniesResponse(message="Company liked successfully", success=True)
+
+@router.post("/unlike-company")
+def unlike_company(
+    request: UnlikeCompanyRequest,
+    db: Session = Depends(database.get_db)
+) -> UnlikedCompaniesResponse:
+    liked_collection = (
+        db.query(database.CompanyCollection)
+        .filter(database.CompanyCollection.collection_name == "Liked Companies")
+        .first()
+    )   
+    
+    # Check if company is in liked collection
+    liked_association = db.query(database.CompanyCollectionAssociation).filter(
+        database.CompanyCollectionAssociation.collection_id == liked_collection.id,
+        database.CompanyCollectionAssociation.company_id == request.company_id
+    ).first()
+    if not liked_association:
+        return UnlikedCompaniesResponse(message="Company not in liked collection", success=False)
+    
+    db.delete(liked_association)
+    db.commit()
+    return UnlikedCompaniesResponse(message="Company unliked successfully", success=True)
+
+
+@router.post("/add-collection-to-collection")
+async def add_collection_to_collection(
+    request: AddAllCompaniesToCollectionRequest,
+    background_tasks: BackgroundTasks,
+    db: Session = Depends(database.get_db),
+    db_async: AsyncSession = Depends(database.get_async_db),
```

Comment:
I just realized I forgot to clean this up while I was in local development: should only be either synchronous or async db operator (in this case async).